Panic importing `vault_mfa_duo`

### Terraform Version
```
Terraform v1.2.8
on linux_amd64
+ provider registry.terraform.io/hashicorp/vault v2.24.1
```

### Affected Resource(s)
Please list the resources as a list, for example:
- vault_mfa_duo

If this issue appears to affect multiple resources, it may be an issue with Terraform's core, so please mention this.

### Terraform Configuration Files
```hcl
resource "vault_mfa_duo" "duo" {
  name            = "duo"
  mount_accessor  = vault_ldap_auth_backend.ldap.accessor
  secret_key      = "xxx"
  integration_key = "xxx"
  api_hostname    = "api-2b5c39f5.duosecurity.com"
}
```


### Panic Output

```
❯ terraform import vault_mfa_duo.duo duo
vault_mfa_duo.duo: Importing from ID "duo"...
vault_mfa_duo.duo: Import prepared!
  Prepared vault_mfa_duo for import
vault_mfa_duo.duo: Refreshing state... [id=duo]
╷
│ Error: Request cancelled
│ 
│ The plugin.(*GRPCProvider).ReadResource request was cancelled.
╵


Stack trace from the terraform-provider-vault_v2.24.1_x4 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0xf1e154]

goroutine 14 [running]:
github.com/hashicorp/terraform-provider-vault/vault.mfaDuoRead(0xc00088e8c0, 0x119aa00, 0xc0003b8280, 0xc00088e8c0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-vault/vault/resource_mfa_duo.go:115 +0x2f4
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc00056def0, 0xc000740c30, 0x119aa00, 0xc0003b8280, 0xc0007e6238, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.9.0/helper/schema/resource.go:460 +0x12e
github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ReadResource(0xc000286c30, 0x13c3450, 0xc0008913b0, 0xc0007ca360, 0xc000286c30, 0xc0008913b0, 0xc0005d1ba0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.9.0/internal/helper/plugin/grpc_provider.go:525 +0x3dd
github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ReadResource_Handler(0x1168220, 0xc000286c30, 0x13c3450, 0xc0008913b0, 0xc0007ca300, 0x0, 0x13c3450, 0xc0008913b0, 0xc00047c6c0, 0x90)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.9.0/internal/tfplugin5/tfplugin5.pb.go:3269 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000360000, 0x13ce238, 0xc000103980, 0xc00092e100, 0xc000692330, 0x19de4b0, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:1082 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc000360000, 0x13ce238, 0xc000103980, 0xc00092e100, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:1405 +0xccf
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0000345b0, 0xc000360000, 0x13ce238, 0xc000103980, 0xc00092e100)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:746 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:744 +0xa5

Error: The terraform-provider-vault_v2.24.1_x4 plugin crashed!
```

--- a/vault/resource_mfa_duo.go
+++ b/vault/resource_mfa_duo.go
@@ -1,10 +1,11 @@
 package vault
 
 import (
-	"fmt"
+	"context"
 	"log"
 	"strings"
 
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 
 	"github.com/hashicorp/terraform-provider-vault/internal/consts"
@@ -13,49 +14,52 @@ import (
 
 func mfaDuoResource() *schema.Resource {
 	return &schema.Resource{
-		Create: mfaDuoWrite,
-		Update: mfaDuoWrite,
-		Delete: mfaDuoDelete,
-		Read:   ReadWrapper(mfaDuoRead),
+		CreateContext: mfaDuoWrite,
+		UpdateContext: mfaDuoWrite,
+		DeleteContext: mfaDuoDelete,
+		ReadContext:   ReadContextWrapper(mfaDuoRead),
 		Importer: &schema.ResourceImporter{
-			State: schema.ImportStatePassthrough,
+			StateContext: schema.ImportStatePassthroughContext,
 		},
 
 		Schema: map[string]*schema.Schema{
-			"name": {
+			consts.FieldName: {
 				Type:         schema.TypeString,
 				Required:     true,
 				Description:  "Name of the MFA method.",
 				ValidateFunc: provider.ValidateNoTrailingSlash,
 			},
 			consts.FieldMountAccessor: {
-				Type:        schema.TypeString,
-				Required:    true,
-				Description: "The mount to tie this method to for use in automatic mappings. The mapping will use the Name field of Aliases associated with this mount as the username in the mapping.",
+				Type:     schema.TypeString,
+				Required: true,
+				Description: "The mount to tie this method to for use in automatic mappings. " +
+					"The mapping will use the Name field of Aliases associated with this mount as " +
+					"the username in the mapping.",
 			},
-			"username_format": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				Description: "A format string for mapping Identity names to MFA method names. Values to substitute should be placed in `{{}}`.",
+			consts.FieldUsernameFormat: {
+				Type:     schema.TypeString,
+				Optional: true,
+				Description: "A format string for mapping Identity names to MFA method names. " +
+					"Values to substitute should be placed in `{{}}`.",
 			},
-			"secret_key": {
+			consts.FieldSecretKey: {
 				Type:        schema.TypeString,
 				Required:    true,
 				Sensitive:   true,
 				Description: "Secret key for Duo.",
 			},
-			"integration_key": {
+			consts.FieldIntegrationKey: {
 				Type:        schema.TypeString,
 				Required:    true,
 				Sensitive:   true,
 				Description: "Integration key for Duo.",
 			},
-			"api_hostname": {
+			consts.FieldAPIHostname: {
 				Type:        schema.TypeString,
 				Required:    true,
 				Description: "API hostname for Duo.",
 			},
-			"push_info": {
+			consts.FieldPushInfo: {
 				Type:        schema.TypeString,
 				Optional:    true,
 				Description: "Push information for Duo.",
@@ -64,13 +68,13 @@ func mfaDuoResource() *schema.Resource {
 	}
 }
 
-func mfaDuoWrite(d *schema.ResourceData, meta interface{}) error {
+func mfaDuoWrite(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	client, e := provider.GetClient(d, meta)
 	if e != nil {
-		return e
+		return diag.FromErr(e)
 	}
 
-	name := d.Get("name").(string)
+	name := d.Get(consts.FieldName).(string)
 
 	data := map[string]interface{}{}
 	mfaDuoUpdateFields(d, data)
@@ -81,55 +85,60 @@ func mfaDuoWrite(d *schema.ResourceData, meta interface{}) error {
 	log.Printf("[DEBUG] Creating mfaDuo %s in Vault", name)
 	_, err := client.Logical().Write(mfaDuoPath(name), data)
 	if err != nil {
-		return fmt.Errorf("error writing to Vault: %s", err)
+		return diag.Errorf("error writing to Vault: %s", err)
 	}
 
-	return mfaDuoRead(d, meta)
+	return mfaDuoRead(ctx, d, meta)
 }
 
-func mfaDuoDelete(d *schema.ResourceData, meta interface{}) error {
+func mfaDuoDelete(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	client, e := provider.GetClient(d, meta)
 	if e != nil {
-		return e
+		return diag.FromErr(e)
 	}
 
-	name := d.Get("name").(string)
+	name := d.Get(consts.FieldName).(string)
 
 	log.Printf("[DEBUG] Deleting mfaDuo %s from Vault", mfaDuoPath(name))
 
 	_, err := client.Logical().Delete(mfaDuoPath(name))
 	if err != nil {
-		return fmt.Errorf("error deleting from Vault: %s", err)
+		return diag.Errorf("error deleting from Vault: %s", err)
 	}
 
 	return nil
 }
 
-func mfaDuoRead(d *schema.ResourceData, meta interface{}) error {
+func mfaDuoRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	client, e := provider.GetClient(d, meta)
 	if e != nil {
-		return e
+		return diag.FromErr(e)
 	}
 
-	name := d.Get("name").(string)
+	name := d.Id()
 
 	resp, err := client.Logical().Read(mfaDuoPath(name))
 	if err != nil {
-		return fmt.Errorf("error reading from Vault: %s", err)
+		return diag.Errorf("error reading from Vault: %s", err)
 	}
 
 	log.Printf("[DEBUG] Read MFA Duo config %q", mfaDuoPath(name))
 
-	d.Set(consts.FieldMountAccessor, resp.Data[consts.FieldMountAccessor])
-	d.Set("username_format", resp.Data["username_format"])
-	d.Set("api_hostname", resp.Data["api_hostname"])
+	if err := d.Set(consts.FieldMountAccessor, resp.Data[consts.FieldMountAccessor]); err != nil {
+		return diag.FromErr(err)
+	}
 
-	// When you push the data up, it's push_info
-	// when vault responds, it's pushinfo :(
-	d.Set("push_info", resp.Data["pushinfo"])
+	if err := d.Set(consts.FieldUsernameFormat, resp.Data[consts.FieldUsernameFormat]); err != nil {
+		return diag.FromErr(err)
+	}
 
-	// secret_key and integration_key, can't read out from the api
-	// So... if it drifts, it drift.
+	if err := d.Set(consts.FieldAPIHostname, resp.Data[consts.FieldAPIHostname]); err != nil {
+		return diag.FromErr(err)
+	}
+
+	if err := d.Set(consts.FieldPushInfo, resp.Data["pushinfo"]); err != nil {
+		return diag.FromErr(err)
+	}
 
 	d.SetId(name)
 
@@ -141,24 +150,24 @@ func mfaDuoUpdateFields(d *schema.ResourceData, data map[string]interface{}) {
 		data[consts.FieldMountAccessor] = v.(string)
 	}
 
-	if v, ok := d.GetOk("username_format"); ok {
-		data["username_format"] = v.(string)
+	if v, ok := d.GetOk(consts.FieldUsernameFormat); ok {
+		data[consts.FieldUsernameFormat] = v.(string)
 	}
 
-	if v, ok := d.GetOk("secret_key"); ok {
-		data["secret_key"] = v.(string)
+	if v, ok := d.GetOk(consts.FieldSecretKey); ok {
+		data[consts.FieldSecretKey] = v.(string)
 	}
 
-	if v, ok := d.GetOk("integration_key"); ok {
-		data["integration_key"] = v.(string)
+	if v, ok := d.GetOk(consts.FieldIntegrationKey); ok {
+		data[consts.FieldIntegrationKey] = v.(string)
 	}
 
-	if v, ok := d.GetOk("api_hostname"); ok {
-		data["api_hostname"] = v.(string)
+	if v, ok := d.GetOk(consts.FieldAPIHostname); ok {
+		data[consts.FieldAPIHostname] = v.(string)
 	}
 
-	if v, ok := d.GetOk("push_info"); ok {
-		data["push_info"] = v.(string)
+	if v, ok := d.GetOk(consts.FieldPushInfo); ok {
+		data[consts.FieldPushInfo] = v.(string)
 	}
 }
 

--- a/vault/resource_mfa_duo.go
+++ b/vault/resource_mfa_duo.go
@@ -118,6 +118,12 @@ func mfaDuoRead(_ context.Context, d *schema.ResourceData, meta interface{}) dia
 	name := d.Id()
 
 	resp, err := client.Logical().Read(mfaDuoPath(name))
+	if resp == nil {
+		log.Printf("[WARN] %q not found, removing from state", mfaDuoPath(name))
+		d.SetId("")
+		return nil
+	}
+
 	if err != nil {
 		return diag.Errorf("error reading from Vault: %s", err)
 	}

--- a/vault/resource_mfa_duo_test.go
+++ b/vault/resource_mfa_duo_test.go
@@ -13,6 +13,8 @@ import (
 func TestMFADuoBasic(t *testing.T) {
 	mfaDuoPath := acctest.RandomWithPrefix("mfa-duo")
 
+	resourceName := "vault_mfa_duo.test"
+
 	resource.Test(t, resource.TestCase{
 		PreCheck:  func() { testutil.TestEntPreCheck(t) },
 		Providers: testProviders,
@@ -20,14 +22,18 @@ func TestMFADuoBasic(t *testing.T) {
 			{
 				Config: testMFADuoConfig(mfaDuoPath),
 				Check: resource.ComposeTestCheckFunc(
-					resource.TestCheckResourceAttr("vault_mfa_duo.test", "name", mfaDuoPath),
-					resource.TestCheckResourceAttr("vault_mfa_duo.test", "secret_key", "8C7THtrIigh2rPZQMbguugt8IUftWhMRCOBzbuyz"),
-					resource.TestCheckResourceAttr("vault_mfa_duo.test", "integration_key", "BIACEUEAXI20BNWTEYXT"),
-					resource.TestCheckResourceAttr("vault_mfa_duo.test", "api_hostname", "api-2b5c39f5.duosecurity.com"),
-					resource.TestCheckResourceAttr("vault_mfa_duo.test", "username_format", "user@example.com"),
-					resource.TestCheckResourceAttr("vault_mfa_duo.test", "push_info", "from=loginortal&domain=example.com"),
+					resource.TestCheckResourceAttr(resourceName, "name", mfaDuoPath),
+					resource.TestCheckResourceAttr(resourceName, "secret_key", "8C7THtrIigh2rPZQMbguugt8IUftWhMRCOBzbuyz"),
+					resource.TestCheckResourceAttr(resourceName, "integration_key", "BIACEUEAXI20BNWTEYXT"),
+					resource.TestCheckResourceAttr(resourceName, "api_hostname", "api-2b5c39f5.duosecurity.com"),
+					resource.TestCheckResourceAttr(resourceName, "username_format", "user@example.com"),
+					resource.TestCheckResourceAttr(resourceName, "push_info", "from=loginortal&domain=example.com"),
 				),
 			},
+			testutil.GetImportTestStep(resourceName, false, nil,
+				"name",
+				"secret_key",
+				"integration_key"),
 		},
 	})
 }

